Brokerages/Zerodha/ZerodhaBrokerage.cs

```diff
@@ -0,0 +1,1352 @@
+
+        private readonly string _tradingSegment;
+        private readonly string _zerodhaProductType;
+
+        private DateTime _lastTradeTickTime;
+        #endregion
+
+
+
+        /// <summary>
+        /// Constructor for brokerage
+        /// </summary>
+        /// <param name="aggregator">data aggregator </param>
+        /// <param name="tradingSegment">trading segment</param>
+        /// <param name="zerodhaProductType">zerodha product type - MIS, CNC or NRML </param>
+        /// <param name="apiKey">api key</param>
+        /// <param name="apiSecret">api secret</param>
+        /// <param name="algorithm">the algorithm instance is required to retrieve account type</param>
+        public ZerodhaBrokerage(string tradingSegment, string zerodhaProductType, string apiKey, string apiSecret, IAlgorithm algorithm, IDataAggregator aggregator)
+            : base("Zerodha")
+        {
+            _tradingSegment = tradingSegment;
+            _zerodhaProductType = zerodhaProductType;
+            _algorithm = algorithm;
+            _aggregator = aggregator;
+            _kite = new Kite(apiKey, apiSecret);
+            _apiKey = apiKey;
+            _accessToken = apiSecret;
+            WebSocket = new ZerodhaWebSocketClientWrapper();
+            _wssUrl += string.Format(CultureInfo.InvariantCulture, "?api_key={0}&access_token={1}", _apiKey, _accessToken);
+            WebSocket.Initialize(_wssUrl);
+            WebSocket.Message += OnMessage;
+            WebSocket.Open += (sender, args) =>
+            {
+                Log.Trace($"ZerodhaBrokerage(): WebSocket.Open. Subscribing");
+                Subscribe(GetSubscribed());
+            };
+            WebSocket.Error += OnError;
+            _symbolMapper = new ZerodhaSymbolMapper(_kite);
+
+            var subscriptionManager = new EventBasedDataQueueHandlerSubscriptionManager();
+            subscriptionManager.SubscribeImpl += (s, t) =>
+            {
+                Subscribe(s);
+                return true;
+            };
+            subscriptionManager.UnsubscribeImpl += (s, t) => Unsubscribe(s);
+            SubscriptionManager = subscriptionManager;
+
+            Log.Trace("Start Zerodha Brokerage");
+        }
+
+        /// <summary>
+        /// Subscribes to the requested symbols (using an individual streaming channel)
+        /// </summary>
+        /// <param name="symbols">The list of symbols to subscribe</param>
+        public void Subscribe(IEnumerable<Symbol> symbols)
+        {
+            if (symbols.Count() <= 0)
+            {
+                return;
+            }
+            foreach (var symbol in symbols)
+            {
+                var market = GetSymbolMarket(symbol);
+                var instrumentToken = _symbolMapper.GetZerodhaInstrumentToken(symbol.ID.Symbol, market);
+                if (instrumentToken == 0)
+                {
+                    Log.Error("Invalid Zerodha Instrument token");
```

Comment:
Should `continue` in this case? Suggest logging the Lean symbol

nit - Log should be of the format `ZerodhaBrokerage.Subscribe():....`

Same comments for `Unsubscribe`